File: backend/app/api/v1/posts.py
from datetime import datetime
from typing import Literal

from fastapi import APIRouter, Depends, Query, Response, status
from sqlalchemy.orm import Session

from app.api.deps import get_current_user, get_optional_user
from app.db.session import get_db
from app.models.post import PostStatus
from app.models.user import User
from app.schemas.comment import CommentCreate, CommentPublic
from app.schemas.post import PostCreate, PostList, PostPublic, PostUpdate
from app.services.comments import create_comment, list_comments
from app.services.posts import create_post, delete_post, get_post_detail, get_post_or_404, list_posts, search_posts, update_post

router = APIRouter(prefix="/posts", tags=["posts"])


@router.post("", response_model=PostPublic, status_code=status.HTTP_201_CREATED)
def create(payload: PostCreate, db: Session = Depends(get_db), current_user: User = Depends(get_current_user)):
    return create_post(db, current_user, payload)


@router.get("", response_model=list[PostList])
def list_all(
    skip: int = Query(0, ge=0),
    limit: int = Query(20, ge=1, le=100),
    status_filter: PostStatus | None = Query(default=None, alias="status"),
    author_id: int | None = Query(default=None),
    sort_by: Literal["created_at", "views", "likes", "comments", "score"] | None = Query(default=None),
    date_from: datetime | None = Query(default=None),
    date_to: datetime | None = Query(default=None),
    category: str | None = Query(default=None),
    db: Session = Depends(get_db),
    current_user: User | None = Depends(get_optional_user),
):
    return list_posts(
        db,
        skip=skip,
        limit=limit,
        status_filter=status_filter.value if status_filter else None,
        user=current_user,
        author_id=author_id,
        sort_by=sort_by,
        date_from=date_from,
        date_to=date_to,
        category=category,
    )


@router.get("/search", response_model=list[PostList])
def search(
    keyword: str = Query(min_length=1),
    skip: int = 0,
    limit: int = Query(20, ge=1, le=100),
    sort_by: Literal["created_at", "views", "likes", "comments", "score"] | None = Query(default=None),
    date_from: datetime | None = Query(default=None),
    date_to: datetime | None = Query(default=None),
    fuzzy: bool = Query(False),
    category: str | None = Query(default=None),
    db: Session = Depends(get_db),
):
    return search_posts(
        db,
        keyword=keyword,
        skip=skip,
        limit=limit,
        sort_by=sort_by,
        date_from=date_from,
        date_to=date_to,
        fuzzy=fuzzy,
        category=category,
    )


@router.get("/{post_id}", response_model=PostPublic)
def detail(post_id: int, db: Session = Depends(get_db)):
    return get_post_detail(db, post_id)


@router.put("/{post_id}", response_model=PostPublic)
def update(post_id: int, payload: PostUpdate, db: Session = Depends(get_db), current_user: User = Depends(get_current_user)):
    post = get_post_or_404(db, post_id)
    return update_post(db, post, current_user, payload)


@router.delete("/{post_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete(post_id: int, db: Session = Depends(get_db), current_user: User = Depends(get_current_user)):
    post = get_post_or_404(db, post_id)
    delete_post(db, post, current_user)
    return Response(status_code=status.HTTP_204_NO_CONTENT)


@router.post("/{post_id}/comments", response_model=CommentPublic, status_code=status.HTTP_201_CREATED)
def comment(post_id: int, payload: CommentCreate, db: Session = Depends(get_db), current_user: User = Depends(get_current_user)):
    post = get_post_or_404(db, post_id)
    return create_comment(db, current_user, payload, post=post)


@router.get("/{post_id}/comments", response_model=list[CommentPublic])
def comments(post_id: int, db: Session = Depends(get_db)):
    get_post_or_404(db, post_id)
    return list_comments(db, post_id)

File: frontend/src/api/posts.ts
import { api } from "./client";
import type { Post, PostCategory, PostStatus } from "../types";

export type PostPayload = {
  title: string;
  summary?: string;
  content: string;
  cover_url?: string;
  tags?: string[];
  category: PostCategory;
  status: PostStatus;
};

export type SortBy = "created_at" | "views" | "likes" | "comments" | "score";

export const TAG_PRESETS = ["Python", "JavaScript", "React", "FastAPI", "PostgreSQL", "Docker", "Linux", "AI", "开源", "教程", "随笔"];

export async function listPosts(params?: {
  keyword?: string;
  status?: PostStatus;
  author_id?: number;
  sort_by?: SortBy;
  date_from?: string;
  date_to?: string;
  fuzzy?: boolean;
  limit?: number;
  category?: PostCategory;
}) {
  const endpoint = params?.keyword ? "/posts/search" : "/posts";
  const query: Record<string, string> = {};
  if (params?.keyword) query.keyword = params.keyword;
  if (params?.status) query.status = params.status;
  if (params?.author_id) query.author_id = String(params.author_id);
  if (params?.sort_by) query.sort_by = params.sort_by;
  if (params?.date_from) query.date_from = params.date_from;
  if (params?.date_to) query.date_to = params.date_to;
  if (params?.fuzzy) query.fuzzy = "true";
  if (params?.limit) query.limit = String(params.limit);
  if (params?.category) query.category = params.category;
  const { data } = await api.get<Post[]>(endpoint, { params: query });
  return data;
}

export async function getPost(id: number) {
  const { data } = await api.get<Post>(`/posts/${id}`);
  return data;
}

export async function createPost(payload: PostPayload) {
  const { data } = await api.post<Post>("/posts", payload);
  return data;
}

export async function updatePost(id: number, payload: Partial<PostPayload>) {
  const { data } = await api.put<Post>(`/posts/${id}`, payload);
  return data;
}

export async function deletePost(id: number) {
  await api.delete(`/posts/${id}`);
}
